lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -44,6 +106,132 @@ SerializedModuleLoader::~SerializedModuleLoader() = default;
+      [&](StringRef dirPath, llvm::function_ref<void(StringRef)> body) {
+        std::error_code errorCode;
+        llvm::vfs::directory_iterator DI = fs.dir_begin(dirPath, errorCode);
+        llvm::vfs::directory_iterator End;
+        for (; !errorCode && DI != End; DI.increment(errorCode))
+          body(DI->path());
+      };
+
+  // Check whether target specific module file exists or not in given directory.
+  // $PATH/{arch}.{extension}
+  auto checkTargetFiles = [&](StringRef path) -> bool {
+    llvm::SmallString<256> scratch;
+    for (auto targetFile : targetFiles) {
+      scratch.clear();
+      llvm::sys::path::append(scratch, path, targetFile);
+      // If {arch}.{extension} exists, consider it's visible. Technically, we
+      // should check the file type, permission, format, etc., but it's too
+      // heavy to do that for each files.
+      if (fs.exists(scratch))
+        return true;
+    }
+    return false;
+  };
+
+  forEachModuleSearchPath(Ctx, [&](StringRef searchPath, SearchPathKind Kind) {
+    switch (Kind) {
+    case SearchPathKind::Import: {
+      // Look for:
+      // $PATH/{name}.swiftmodule/{arch}.{extension} ; Or
+      // $PATH/{name}.{extension}
+      forEachDirectoryEntryPath(searchPath, [&](StringRef path) {
+        auto pathExt = llvm::sys::path::extension(path);
+        if (pathExt != moduleSuffix && pathExt != suffix)
+          return;
+
+        auto stat = fs.status(path);
+        if (!stat)
+          return;
+        if (pathExt == moduleSuffix && stat->isDirectory()) {
+          if (!checkTargetFiles(path))
+            return;
+        } else if (pathExt != suffix || stat->isDirectory()) {
+          return;
+        }
+        // Extract module name.
+        auto name = llvm::sys::path::filename(path).drop_back(pathExt.size());
+        names.push_back(Ctx.getIdentifier(name));
+      });
+      return None;
+    }
+    case SearchPathKind::RuntimeLibrary: {
+      // Look for:
+      // (Darwin OS) $PATH/{name}.swiftmodule/{arch}.{extension}
+      // (Other OS)  $PATH/{name}.{extension}
+      bool requireTargetSpecificModule = Ctx.LangOpts.Target.isOSDarwin();
+      forEachDirectoryEntryPath(searchPath, [&](StringRef path) {
+        auto pathExt = llvm::sys::path::extension(path);
+        if (requireTargetSpecificModule) {
+          if (pathExt != moduleSuffix)
+            return;
+          if (!checkTargetFiles(path))
+            return;
+        } else {
+          if (suffix != pathExt)
+            return;
+          auto stat = fs.status(path);
+          if (!stat || stat->isDirectory())
+            return;
+        }
+        // Extract module name.
+        auto name = llvm::sys::path::filename(path).drop_back(pathExt.size());
+        names.push_back(Ctx.getIdentifier(name));
+      });
+      return None;
+    }
+    case SearchPathKind::Framework: {
+      // Look for:
+      // $PATH/{name}.framework/Modules/{name}.swiftmodule/{arch}.{extension}
```

Comment:
Not sure we need to handle frameworks practically. If framework always have Clang module (`module.modulemap`) with corresponding module name, `ClangImporter` can find them.
@jrose-apple Is pure Swift only *framework* possible?